scripts/build-tools.sh

```diff
@@ -5,69 +5,126 @@
 # fail immediately on any errors
 set -e
 
-
 print_usage()
 {
         cat <<EOFUSAGE
-usage: $0 [-t|-f]
-       [-t] Build test topologies
-       [-f] Build fuzzer"
+usage: $0 [-c|-f|-h|-l|-p|-t|-T]
+       -c Rebuild ctl
+       -f Rebuild fuzzer
+       -h Display help
+       -l Rebuild logger
+       -p Rebuild probes
+       -t Rebuild test topologies
+       -T Rebuild topologies
+       -C No build, only CMake re-configuration
 EOFUSAGE
 }
 
-build_tools()
+# generate Makefiles
+prepare_build()
 {
-        cd "$SOF_REPO/tools"
-        rm -rf build_tools
-        mkdir build_tools
-        cd build_tools
+        rm -rf "$BUILD_TOOLS_DIR"
+        mkdir "$BUILD_TOOLS_DIR"
+
+        cd "$BUILD_TOOLS_DIR"
         cmake ..
-        make -j "$NO_PROCESSORS"
+
+        mkdir "$BUILD_TOOLS_DIR/fuzzer"
+        cd "$BUILD_TOOLS_DIR/fuzzer"
+        cmake ../../fuzzer
 }
 
-build_test()
+make_tool()
 {
-        cd "$SOF_REPO/tools/build_tools"
-        make tests -j "$NO_PROCESSORS"
+        # if no argument provided, all the tools will be built. Empty tool is
+        # okay.
+        tool=$1
+        make -C "$BUILD_TOOLS_DIR" -j "$NO_PROCESSORS" $tool
 }
 
-build_fuzzer()
+make_fuzzer()
 {
-        cd "$SOF_REPO/tools/build_tools"
-        mkdir build_fuzzer
-        cd build_fuzzer
-        cmake ../../fuzzer
-        make -j "$NO_PROCESSORS"
+        make -C "$BUILD_TOOLS_DIR/fuzzer" -j "$NO_PROCESSORS"
+}
+
+print_build_info()
+{
+       cat <<EOFUSAGE
+
+Build commands for respective tools:
+        ctl:        make -C "$BUILD_TOOLS_DIR" sof-ctl
+        logger:     make -C "$BUILD_TOOLS_DIR" sof-logger
+        probes:     make -C "$BUILD_TOOLS_DIR" sof-probes
+        tests:      make -C "$BUILD_TOOLS_DIR" tests
+        topologies: make -C "$BUILD_TOOLS_DIR" topologies
+        fuzzer:     make -C "$BUILD_TOOLS_DIR/fuzzer"
+EOFUSAGE
 }
 
 main()
 {
-        local DO_BUILD_TEST DO_BUILD_FUZZER SCRIPT_DIR SOF_REPO
+        local DO_BUILD_ctl DO_BUILD_fuzzer DO_BUILD_logger DO_BUILD_probes \
+                DO_BUILD_test DO_BUILD_topology SCRIPT_DIR SOF_REPO CMAKE_ONLY \
+                BUILD_ALL
         SCRIPT_DIR=$(cd "$(dirname "$0")" && pwd)
         SOF_REPO=$(dirname "$SCRIPT_DIR")
+        BUILD_TOOLS_DIR="$SOF_REPO"/tools/build_tools
         : ${NO_PROCESSORS:=$(nproc)}
+        BUILD_ALL=false
+
+        if [ $# -eq 0 ]; then
+                BUILD_ALL=true
+        fi
```

Comment:
Can you move this after parsing options, so it's ready for adding a flag that affects all builds in the future?

Reply:
if we move this zero arg test below getopts, we shift everything by then and it will always be 0. Everything gets parsed and no way to know if any arguments were provided(build specific things) or none were provided(build everything).

Reply:
Yes, and I was trying to avoid not having to set `BUILD_ALL` with each option provided.